README.md
# Elliptic curve cryptography (ECC) in Dart

Elliptic curve cryptography lib for EOS based blockchain in Dart lang.

[![Build Status](https://travis-ci.org/bcrumbs/snaxdart_ecc.svg?branch=master)](https://travis-ci.org/bcrumbs/snaxdart_ecc)

## Usage

A simple usage example:

```dart
import 'package:snaxdart_ecc/snaxdart_ecc.dart';

main() {
  // Construct the SNAX private key from string
  SNAXPrivateKey privateKey = SNAXPrivateKey.fromString(
      '5KQwrPbwdL6PhXujxW37FSSQZ1JiwsST4cqQzDeyXtP79zkvFD3');

  // Get the related SNAX public key
  SNAXPublicKey publicKey = privateKey.toSNAXPublicKey();
  // Print the SNAX public key
  print(publicKey.toString());

  // Going to sign the data
  String data = 'data';

  // Sign
  SNAXSignature signature = privateKey.signString(data);
  // Print the SNAX signature
  print(signature.toString());

  // Verify the data using the signature
  signature.verify(data, publicKey);
}
```

## Features and bugs

Please file feature requests and bugs at the [issue tracker][tracker].

## References

eosjs-ecc: https://github.com/EOSIO/eosjs-ecc

[tracker]: https://github.com/primes-network/snaxdart_ecc/issues


CHANGELOG.md
## 1.0.0

- migrate to Snax

## 0.3.2

- fix signature bug

## 0.3.1

- sign method will sign Uint8List
- add signString method which signs String

## 0.3.0

- Support canonical signature

## 0.2.2

- Healthy change

## 0.2.1

- Add CHANGELOG.md

## 0.2.0

- SNAXPublicKey, SNAXPrivateKey, SNAXSignature
- Sign the data
- Verify the signature

## 0.1.0

- Initial version, created by Stagehand


test/signature_test.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:snaxdart_ecc/snaxdart_ecc.dart';
import 'package:test/test.dart';
import 'package:crypto/crypto.dart';

void main() {
  group('SNAX signature tests', () {
    test('Construct SNAX signature from string', () {
      String sigStr =
          'SIG_K1_Kg417TSLuhzSpU2bGa21kD1UNaTfAZSCcKmKpZ6fnx3Nqu22gzG3ND4Twur7bzX8oS1J91JvV4rMJcFycGqFBSaY2SJcEQ';
      SNAXSignature signature = SNAXSignature.fromString(sigStr);
      print(signature);

      expect(sigStr, signature.toString());
    });

    test('Sign the hash using private key', () {
      SNAXPrivateKey privateKey = SNAXPrivateKey.fromString(
          '5KQwrPbwdL6PhXujxW37FSSQZ1JiwsST4cqQzDeyXtP79zkvFD3');
      SNAXPublicKey publicKey = privateKey.toSNAXPublicKey();
      String expectedSig =
          'SIG_K1_Kg417TSLuhzSpU2bGa21kD1UNaTfAZSCcKmKpZ6fnx3Nqu22gzG3ND4Twur7bzX8oS1J91JvV4rMJcFycGqFBSaY2SJcEQ';

      String data = 'data';
      Uint8List hashData = sha256.convert(utf8.encode(data)).bytes;
      SNAXSignature signature = privateKey.signHash(hashData);
      SNAXSignature signature2 = privateKey.signString(data);

      print(signature.toString());
      expect(expectedSig, signature.toString());
      expect(true, signature.verifyHash(hashData, publicKey));
      expect(true, signature2.verifyHash(hashData, publicKey));

      expect(true, signature.verify(data, publicKey));
      expect(true, signature2.verify(data, publicKey));
    });

    test('Sign the hash using private key', () {
      SNAXPrivateKey privateKey = SNAXPrivateKey.fromString(
          '5HxT6prWB8VuXkoAaX3eby8bWjquMtCvGuakhC8tGEiPSHfsQLR');
      SNAXPublicKey publicKey = privateKey.toSNAXPublicKey();
      String expectedSig =
          'SIG_K1_Kdfe9wknSAKBmgwb3L53CG8KosoHhZ69oVEJrrH5YuWx4JVcJdn1ZV3MU25AVho4mPbeSKW79DVTBAAWj7zGbHTByF1JXU';

      List<int> l = [
        244,
        163,
        240,
        75,
        174,
        150,
        233,
        185,
        227,
        66,
        27,
        130,
        230,
        139,
        102,
        112,
        128,
        38,
        78,
        233,
        105,
        59,
        61,
        11,
        25,
        221,
        42,
        109,
        80,
        184,
        174,
        201
      ];
      Uint8List hashData = Uint8List.fromList(l);
      SNAXSignature signature = privateKey.signHash(hashData);

      expect(expectedSig, signature.toString());
      print(signature.toString());
      expect(true, signature.verifyHash(hashData, publicKey));
    });

    test('Sign the hash using private key', () {
      SNAXPrivateKey privateKey = SNAXPrivateKey.fromString(
          '5J9b3xMkbvcT6gYv2EpQ8FD4ZBjgypuNKwE1jxkd7Wd1DYzhk88');
      SNAXPublicKey publicKey = privateKey.toSNAXPublicKey();
      String expectedSig =
          'SIG_K1_KWfDGxwogny1PUiBAYTfKwPsCSNvM7zWgmXyChdYayZFfyPjddpBUYVdJTq1PjC3PRXADRsqWVU1N2SMQivBDqA7AaRzmB';

      List<int> l = [
        136,
        139,
        63,
        11,
        114,
        68,
        227,
        116,
        92,
        61,
        64,
        121,
        147,
        210,
        233,
        25,
        74,
        164,
        140,
        112,
        45,
        5,
        254,
        165,
        208,
        158,
        53,
        212,
        128,
        190,
        153,
        142
      ];
      Uint8List hashData = Uint8List.fromList(l);
      SNAXSignature signature = privateKey.signHash(hashData);

      expect(expectedSig, signature.toString());
      print(signature.toString());
      expect(true, signature.verifyHash(hashData, publicKey));
    });
  });
}


test/eosdart_ecc_test.dart
import 'package:snaxdart_ecc/snaxdart_ecc.dart';
import 'package:test/test.dart';

void main() {
  group('SNAX Key tests', () {
    test('Construct SNAX public key from string', () {
      SNAXPublicKey publicKey = SNAXPublicKey.fromString(
          'SNAX8Qi58kbERkTJC7A4gabxYU4SbrAxStJHacoke4sf6AvJyEDZXj');
      print(publicKey);

      expect('SNAX8Qi58kbERkTJC7A4gabxYU4SbrAxStJHacoke4sf6AvJyEDZXj',
          publicKey.toString());
    });

    test('Construct SNAX public key from string PUB_K1 format', () {
      SNAXPublicKey publicKey = SNAXPublicKey.fromString(
          'PUB_K1_859gxfnXyUriMgUeThh1fWv3oqcpLFyHa3TfFYC4PK2Ht7beeX');
      print(publicKey);
    });

    test('Construct SNAX private key from string', () {
      // common private key
      SNAXPrivateKey privateKey = SNAXPrivateKey.fromString(
          '5J9b3xMkbvcT6gYv2EpQ8FD4ZBjgypuNKwE1jxkd7Wd1DYzhk88');
      expect('SNAX8Qi58kbERkTJC7A4gabxYU4SbrAxStJHacoke4sf6AvJyEDZXj',
          privateKey.toSNAXPublicKey().toString());
      expect('5J9b3xMkbvcT6gYv2EpQ8FD4ZBjgypuNKwE1jxkd7Wd1DYzhk88',
          privateKey.toString());
    });

    test('Invalid SNAX private key', () {
      try {
        SNAXPrivateKey.fromString(
            '5KYZdUEo39z3FPrtuX2QbbwGnNP5zTd7yyr2SC1j299sBCnWjsm');
        fail('Should be invalid private key');
      } on InvalidKey {} catch (e) {
        fail('Should throw InvalidKey exception');
      }
    });

    test('Construct random SNAX private key from seed', () {
      SNAXPrivateKey privateKey = SNAXPrivateKey.fromSeed('abc');
      print(privateKey);
      print(privateKey.toSNAXPublicKey());

      SNAXPrivateKey privateKey2 =
          SNAXPrivateKey.fromString(privateKey.toString());
      expect(privateKey.toSNAXPublicKey().toString(),
          privateKey2.toSNAXPublicKey().toString());
    });

    test('Construct random SNAX private key', () {
      SNAXPrivateKey privateKey = SNAXPrivateKey.fromRandom();

      print(privateKey);
      print(privateKey.toSNAXPublicKey());

      SNAXPrivateKey privateKey2 =
          SNAXPrivateKey.fromString(privateKey.toString());
      expect(privateKey.toSNAXPublicKey().toString(),
          privateKey2.toSNAXPublicKey().toString());
    });

    test('Construct SNAX private key from string in PVT format', () {
      // PVT private key
      SNAXPrivateKey privateKey2 = SNAXPrivateKey.fromString(
          'PVT_K1_2jH3nnhxhR3zPUcsKaWWZC9ZmZAnKm3GAnFD1xynGJE1Znuvjd');
      print(privateKey2);
    });

    test('Construct SNAX private key from string with compress flag', () {
      // Compressed private key
      SNAXPrivateKey privateKey3 = SNAXPrivateKey.fromString(
          'L5TCkLizyYqjvKSy6jg1XM3Lc4uTDwwvHS2BYatyXSyoS8T5kC2z');
      print(privateKey3);
    });
  });
}


lib/src/key_base.dart
import 'dart:typed_data';
import 'dart:convert';

import 'package:bs58check/bs58check.dart';
import 'package:crypto/crypto.dart';
import 'package:pointycastle/digests/ripemd160.dart';
import 'package:pointycastle/src/utils.dart';
import "package:pointycastle/ecc/curves/secp256k1.dart";

import './exception.dart';

/// abstract SNAX Key
abstract class SNAXKey {
  static final String SHA256X2 = 'sha256x2';
  static final int VERSION = 0x80;
  static final ECCurve_secp256k1 secp256k1 = ECCurve_secp256k1();

  String keyType;

  /// Decode key from string format
  static Uint8List decodeKey(String keyStr, [String keyType]) {
    Uint8List buffer = base58.decode(keyStr);

    Uint8List checksum = buffer.sublist(buffer.length - 4, buffer.length);
    Uint8List key = buffer.sublist(0, buffer.length - 4);

    Uint8List newChecksum;
    if (keyType == SHA256X2) {
      newChecksum = sha256x2(key).sublist(0, 4);
    } else {
      Uint8List check = key;
      if (keyType != null) {
        check = concat(key, utf8.encode(keyType));
      }
      newChecksum = RIPEMD160Digest().process(check).sublist(0, 4);
    }
    if (decodeBigInt(checksum) != decodeBigInt(newChecksum)) {
      throw InvalidKey("checksum error");
    }
    return key;
  }

  /// Encode key to string format using base58 encoding
  static String encodeKey(Uint8List key, [String keyType]) {
    if (keyType == SHA256X2) {
      Uint8List checksum = sha256x2(key).sublist(0, 4);
      return base58.encode(concat(key, checksum));
    }

    Uint8List keyBuffer = key;
    if (keyType != null) {
      keyBuffer = concat(key, utf8.encode(keyType));
    }
    Uint8List checksum = RIPEMD160Digest().process(keyBuffer).sublist(0, 4);
    return base58.encode(concat(key, checksum));
  }

  /// Do SHA256 hash twice on the given data
  static Uint8List sha256x2(Uint8List data) {
    Digest d1 = sha256.convert(data);
    Digest d2 = sha256.convert(d1.bytes);
    return d2.bytes;
  }

  static Uint8List concat(Uint8List p1, Uint8List p2) {
    List<int> keyList = p1.toList();
    keyList.addAll(p2);
    return Uint8List.fromList(keyList);
  }

  static List<int> toSigned(Uint8List bytes) {
    List<int> result = List();
    for (int i = 0; i < bytes.length; i++) {
      int v = bytes[i].toSigned(8);
      //TODO I don't know why, just guess...
      if (i == 0 && v < 0) {
        result.add(0);
      }
      result.add(v);
    }
    return result;
  }
}


lib/src/signature.dart
import 'dart:typed_data';
import 'dart:convert';

import 'package:crypto/crypto.dart';
import "package:pointycastle/api.dart" show PublicKeyParameter;
import 'package:pointycastle/ecc/api.dart'
    show ECPublicKey, ECSignature, ECPoint;
import "package:pointycastle/signers/ecdsa_signer.dart";
import 'package:pointycastle/macs/hmac.dart';
import "package:pointycastle/digests/sha256.dart";
import 'package:pointycastle/src/utils.dart';

import './exception.dart';
import './key.dart';
import './key_base.dart';

/// SNAX Signature
class SNAXSignature extends SNAXKey {
  int i;
  ECSignature ecSig;

  /// Default constructor from i, r, s
  SNAXSignature(this.i, BigInt r, BigInt s) {
    this.keyType = 'K1';
    this.ecSig = ECSignature(r, s);
  }

  /// Construct SNAX signature from buffer
  SNAXSignature.fromBuffer(Uint8List buffer, String keyType) {
    this.keyType = keyType;

    if (buffer.lengthInBytes != 65) {
      throw InvalidKey(
          'Invalid signature length, got: ${buffer.lengthInBytes}');
    }

    i = buffer.first;

    if (i - 27 != i - 27 & 7) {
      throw InvalidKey('Invalid signature parameter');
    }

    BigInt r = decodeBigInt(buffer.sublist(1, 33));
    BigInt s = decodeBigInt(buffer.sublist(33, 65));
    this.ecSig = ECSignature(r, s);
  }

  /// Construct SNAX signature from string
  factory SNAXSignature.fromString(String signatureStr) {
    RegExp sigRegex = RegExp(r"^SIG_([A-Za-z0-9]+)_([A-Za-z0-9]+)",
        caseSensitive: true, multiLine: false);
    Iterable<Match> match = sigRegex.allMatches(signatureStr);

    if (match.length == 1) {
      Match m = match.first;
      String keyType = m.group(1);
      Uint8List key = SNAXKey.decodeKey(m.group(2), keyType);
      return SNAXSignature.fromBuffer(key, keyType);
    }

    throw InvalidKey("Invalid SNAX signature");
  }

  /// Verify the signature of the string data
  bool verify(String data, SNAXPublicKey publicKey) {
    Digest d = sha256.convert(utf8.encode(data));

    return verifyHash(d.bytes, publicKey);
  }

  /// Verify the signature from in SHA256 hashed data
  bool verifyHash(Uint8List sha256Data, SNAXPublicKey publicKey) {
    ECPoint q = publicKey.q;
    final signer = ECDSASigner(null, HMac(SHA256Digest(), 64));
    signer.init(false, PublicKeyParameter(ECPublicKey(q, SNAXKey.secp256k1)));

    return signer.verifySignature(sha256Data, this.ecSig);
  }

  String toString() {
    List<int> b = List();
    b.add(i);
    b.addAll(encodeBigInt(this.ecSig.r));
    b.addAll(encodeBigInt(this.ecSig.s));

    Uint8List buffer = Uint8List.fromList(b);
    return 'SIG_${keyType}_${SNAXKey.encodeKey(buffer, keyType)}';
  }

  /// ECSignature to DER format bytes
  static Uint8List ecSigToDER(ECSignature ecSig) {
    List<int> r = SNAXKey.toSigned(encodeBigInt(ecSig.r));
    List<int> s = SNAXKey.toSigned(encodeBigInt(ecSig.s));

    List<int> b = List();
    b.add(0x02);
    b.add(r.length);
    b.addAll(r);

    b.add(0x02);
    b.add(s.length);
    b.addAll(s);

    b.insert(0, b.length);
    b.insert(0, 0x30);

    return Uint8List.fromList(b);
  }

  /// Find the public key recovery factor
  static int calcPubKeyRecoveryParam(
      BigInt e, ECSignature ecSig, SNAXPublicKey publicKey) {
    for (int i = 0; i < 4; i++) {
      ECPoint Qprime = recoverPubKey(e, ecSig, i);
      if (Qprime == publicKey.q) {
        return i;
      }
    }
    throw 'Unable to find valid recovery factor';
  }

  /// Recovery SNAX public key from ECSignature
  static ECPoint recoverPubKey(BigInt e, ECSignature ecSig, int i) {
    BigInt n = SNAXKey.secp256k1.n;
    ECPoint G = SNAXKey.secp256k1.G;

    BigInt r = ecSig.r;
    BigInt s = ecSig.s;

    // A set LSB signifies that the y-coordinate is odd
    int isYOdd = i & 1;

    // The more significant bit specifies whether we should use the
    // first or second candidate key.
    int isSecondKey = i >> 1;

    // 1.1 Let x = r + jn
    BigInt x = isSecondKey > 0 ? r + n : r;
    ECPoint R = SNAXKey.secp256k1.curve.decompressPoint(isYOdd, x);
    ECPoint nR = R * n;
    if (!nR.isInfinity) {
      throw 'nR is not a valid curve point';
    }

    BigInt eNeg = (-e) % n;
    BigInt rInv = r.modInverse(n);

    ECPoint Q = (R * s + G * eNeg) * rInv;
    return Q;
  }
}


lib/src/key.dart
import 'dart:typed_data';
import 'dart:convert';
import 'dart:math';

import 'package:crypto/crypto.dart';
import 'package:pointycastle/src/utils.dart';
import 'package:pointycastle/ecc/api.dart' show ECSignature, ECPoint;

import './exception.dart';
import './key_base.dart';
import './signature.dart';

/// SNAX Public Key
class SNAXPublicKey extends SNAXKey {
  ECPoint q;

  /// Construct SNAX public key from buffer
  SNAXPublicKey.fromPoint(this.q);

  /// Construct SNAX public key from string
  factory SNAXPublicKey.fromString(String keyStr) {
    RegExp publicRegex = RegExp(r"^PUB_([A-Za-z0-9]+)_([A-Za-z0-9]+)",
        caseSensitive: true, multiLine: false);
    Iterable<Match> match = publicRegex.allMatches(keyStr);

    if (match.isEmpty) {
      RegExp snaxRegex =
          RegExp(r"^SNAX", caseSensitive: true, multiLine: false);
      if (!snaxRegex.hasMatch(keyStr)) {
        throw InvalidKey("No leading SNAX");
      }
      String publicKeyStr = keyStr.substring(4);
      Uint8List buffer = SNAXKey.decodeKey(publicKeyStr);
      return SNAXPublicKey.fromBuffer(buffer);
    } else if (match.length == 1) {
      Match m = match.first;
      String keyType = m.group(1);
      Uint8List buffer = SNAXKey.decodeKey(m.group(2), keyType);
      return SNAXPublicKey.fromBuffer(buffer);
    } else {
      throw InvalidKey('Invalid public key format');
    }
  }

  factory SNAXPublicKey.fromBuffer(Uint8List buffer) {
    ECPoint point = SNAXKey.secp256k1.curve.decodePoint(buffer);
    return SNAXPublicKey.fromPoint(point);
  }

  Uint8List toBuffer() {
    // always compressed
    return q.getEncoded(true);
  }

  String toString() {
    return 'SNAX' + SNAXKey.encodeKey(this.toBuffer(), keyType);
  }
}

/// SNAX Private Key
class SNAXPrivateKey extends SNAXKey {
  Uint8List d;
  String format;

  BigInt _r;
  BigInt _s;

  /// Constructor SNAX private key from the key buffer itself
  SNAXPrivateKey.fromBuffer(this.d);

  /// Construct the private key from string
  /// It can come from WIF format for PVT format
  SNAXPrivateKey.fromString(String keyStr) {
    RegExp privateRegex = RegExp(r"^PVT_([A-Za-z0-9]+)_([A-Za-z0-9]+)",
        caseSensitive: true, multiLine: false);
    Iterable<Match> match = privateRegex.allMatches(keyStr);

    if (match.isEmpty) {
      format = 'WIF';
      keyType = 'K1';
      // WIF
      Uint8List keyWLeadingVersion =
          SNAXKey.decodeKey(keyStr, SNAXKey.SHA256X2);
      int version = keyWLeadingVersion.first;
      if (SNAXKey.VERSION != version) {
        throw InvalidKey("version mismatch");
      }

      d = keyWLeadingVersion.sublist(1, keyWLeadingVersion.length);
      if (d.lengthInBytes == 33 && d.elementAt(32) == 1) {
        // remove compression flag
        d = d.sublist(0, 32);
      }

      if (d.lengthInBytes != 32) {
        throw InvalidKey('Expecting 32 bytes, got ${d.length}');
      }
    } else if (match.length == 1) {
      format = 'PVT';
      Match m = match.first;
      keyType = m.group(1);
      d = SNAXKey.decodeKey(m.group(2), keyType);
    } else {
      throw InvalidKey('Invalid Private Key format');
    }
  }

  /// Generate SNAX private key from seed. Please note: This is not random!
  /// For the given seed, the generated key would always be the same
  factory SNAXPrivateKey.fromSeed(String seed) {
    Digest s = sha256.convert(utf8.encode(seed));
    return SNAXPrivateKey.fromBuffer(s.bytes);
  }

  /// Generate the random SNAX private key
  factory SNAXPrivateKey.fromRandom() {
    final int randomLimit = 1 << 32;
    Random randomGenerator;
    try {
      randomGenerator = Random.secure();
    } catch (e) {
      randomGenerator = new Random();
    }

    int randomInt1 = randomGenerator.nextInt(randomLimit);
    Uint8List entropy1 = encodeBigInt(BigInt.from(randomInt1));

    int randomInt2 = randomGenerator.nextInt(randomLimit);
    Uint8List entropy2 = encodeBigInt(BigInt.from(randomInt2));

    int randomInt3 = randomGenerator.nextInt(randomLimit);
    Uint8List entropy3 = encodeBigInt(BigInt.from(randomInt3));

    List<int> entropy = entropy1.toList();
    entropy.addAll(entropy2);
    entropy.addAll(entropy3);
    Uint8List randomKey = Uint8List.fromList(entropy);
    Digest d = sha256.convert(randomKey);
    return SNAXPrivateKey.fromBuffer(d.bytes);
  }

  /// Check if the private key is WIF format
  bool isWIF() => this.format == 'WIF';

  /// Get the public key string from this private key
  SNAXPublicKey toSNAXPublicKey() {
    BigInt privateKeyNum = decodeBigInt(this.d);
    ECPoint ecPoint = SNAXKey.secp256k1.G * privateKeyNum;

    return SNAXPublicKey.fromPoint(ecPoint);
  }

  /// Sign the bytes data using the private key
  SNAXSignature sign(Uint8List data) {
    Digest d = sha256.convert(data);
    return signHash(d.bytes);
  }

  /// Sign the string data using the private key
  SNAXSignature signString(String data) {
    return sign(utf8.encode(data));
  }

  /// Sign the SHA256 hashed data using the private key
  SNAXSignature signHash(Uint8List sha256Data) {
    int nonce = 0;
    BigInt n = SNAXKey.secp256k1.n;
    BigInt e = decodeBigInt(sha256Data);

    while (true) {
      _deterministicGenerateK(sha256Data, this.d, e, nonce++);
      var N_OVER_TWO = n >> 1;
      if (_s.compareTo(N_OVER_TWO) > 0) {
        _s = n - _s;
      }
      ECSignature sig = ECSignature(_r, _s);

      Uint8List der = SNAXSignature.ecSigToDER(sig);

      int lenR = der.elementAt(3);
      int lenS = der.elementAt(5 + lenR);
      if (lenR == 32 && lenS == 32) {
        int i = SNAXSignature.calcPubKeyRecoveryParam(
            decodeBigInt(sha256Data), sig, this.toSNAXPublicKey());
        i += 4; // compressed
        i += 27; // compact  //  24 or 27 :( forcing odd-y 2nd key candidate)
        return SNAXSignature(i, sig.r, sig.s);
      }
    }
  }

  String toString() {
    List<int> version = List<int>();
    version.add(SNAXKey.VERSION);
    Uint8List keyWLeadingVersion =
        SNAXKey.concat(Uint8List.fromList(version), this.d);

    return SNAXKey.encodeKey(keyWLeadingVersion, SNAXKey.SHA256X2);
  }

  BigInt _deterministicGenerateK(
      Uint8List hash, Uint8List x, BigInt e, int nonce) {
    List<int> newHash = hash;
    if (nonce > 0) {
      List<int> addition = Uint8List(nonce);
      List<int> data = List.from(hash)..addAll(addition);
      newHash = sha256.convert(data).bytes;
    }

    // Step B
    Uint8List v = Uint8List(32);
    for (int i = 0; i < v.lengthInBytes; i++) {
      v[i] = 1;
    }

    // Step C
    Uint8List k = Uint8List(32);

    // Step D
    List<int> d1 = List.from(v)
      ..add(0)
      ..addAll(x)
      ..addAll(newHash);

    Hmac hMacSha256 = new Hmac(sha256, k); // HMAC-SHA256
    k = hMacSha256.convert(d1).bytes;

    // Step E
    hMacSha256 = new Hmac(sha256, k); // HMAC-SHA256
    v = hMacSha256.convert(v).bytes;

    // Step F
    List<int> d2 = List.from(v)
      ..add(1)
      ..addAll(x)
      ..addAll(newHash);

    k = hMacSha256.convert(d2).bytes;

    // Step G
    hMacSha256 = new Hmac(sha256, k); // HMAC-SHA256
    v = hMacSha256.convert(v).bytes;
    // Step H1/H2a, again, ignored as tlen === qlen (256 bit)
    // Step H2b again
    v = hMacSha256.convert(v).bytes;

    BigInt T = decodeBigInt(v);
    // Step H3, repeat until T is within the interval [1, n - 1]
    while (T.sign <= 0 ||
        T.compareTo(SNAXKey.secp256k1.n) >= 0 ||
        !_checkSig(e, newHash, T)) {
      List<int> d3 = List.from(v)..add(0);
      k = hMacSha256.convert(d3).bytes;
      hMacSha256 = new Hmac(sha256, k); // HMAC-SHA256
      v = hMacSha256.convert(v).bytes;
      // Step H1/H2a, again, ignored as tlen === qlen (256 bit)
      // Step H2b again
      v = hMacSha256.convert(v).bytes;

      T = decodeBigInt(v);
    }
    return T;
  }

  bool _checkSig(BigInt e, Uint8List hash, BigInt k) {
    BigInt n = SNAXKey.secp256k1.n;
    ECPoint Q = SNAXKey.secp256k1.G * k;

    if (Q.isInfinity) {
      return false;
    }

    _r = Q.x.toBigInteger() % n;
    if (_r.sign == 0) {
      return false;
    }

    _s = k.modInverse(SNAXKey.secp256k1.n) * (e + decodeBigInt(d) * _r) % n;
    if (_s.sign == 0) {
      return false;
    }

    return true;
  }
}
